setup.py

```diff
@@ -1,84 +1,30 @@
 import os
 import platform
-import sys
 
 from setuptools import Extension
 from setuptools import setup
-from setuptools.command.build_ext import build_ext
-from setuptools.errors import CCompilerError
-from setuptools.errors import ExecError
-from setuptools.errors import PlatformError
+from setuptools_rust import RustExtension
 
-ext_modules = [Extension("markupsafe._speedups", ["src/markupsafe/_speedups.c"])]
-
-
-class BuildFailed(Exception):
-    pass
-
-
-class ve_build_ext(build_ext):
-    """This class allows C extension building to fail."""
-
-    def run(self):
-        try:
-            super().run()
-        except PlatformError as e:
-            raise BuildFailed() from e
-
-    def build_extension(self, ext):
-        try:
-            super().build_extension(ext)
-        except (CCompilerError, ExecError, PlatformError) as e:
-            raise BuildFailed() from e
-        except ValueError as e:
-            # this can happen on Windows 64 bit, see Python issue 7511
-            if "'path'" in str(sys.exc_info()[1]):  # works with Python 2 and 3
-                raise BuildFailed() from e
-            raise
-
-
-def run_setup(with_binary):
-    setup(
-        cmdclass={"build_ext": ve_build_ext},
-        ext_modules=ext_modules if with_binary else [],
-    )
-
-
-def show_message(*lines):
-    print("=" * 74)
-    for line in lines:
-        print(line)
-    print("=" * 74)
-
-
-supports_speedups = platform.python_implementation() not in {
+if platform.python_implementation() not in {
     "PyPy",
     "Jython",
     "GraalVM",
```

Comment:
FWIW we support `PyPy` and `GraalVM`, they _should_ just work in PyO3 for you (or it's a bug for us to resolve).

Reply:
The reason we skip building for them is because the C extension version turned out to be slower than the pure Python version. Do you have any insight into whether this is the case for Rust? My understanding is that those implementations need to emulate Python's C API, which ends up being slower than the speedups their interpreters have.

Reply:
Ah, fair enough. Yes, Rust extensions are built atop the C API and rely on the same emulation / backdoors from their JITs.